src/tests/ftest/io/nvme_io.py

```diff
@@ -67,6 +69,20 @@ def setUp(self):
         self.hostlist = self.params.get("servers", '/run/hosts/*')
         self.hostfile = write_host_file.write_host_file(self.hostlist,
                                                         self.workdir)
+        self.hostlist_clients = self.params.get("clients", '/run/hosts/*')
```

Comment:
It's a shame the apricot.TestWithServers class couldn't have the argument parsing section of it's setUp() be defined in one class method and the hostfile writing and agent/server startup in another class method, such that this class could inherit TestWithServers, and then override setUp() with:

def setUp(self):
    super(NvmeIo, self).setup_params()
    self.nvme_parameter = self.params.get("bdev_class", '/server_config/')
    if self.nvme_parameter == "nvme":
        try:
            spdk.nvme(self.hostlist_servers, "setup")
        except spdk.SpdkFailed as error:
            self.fail("Error setting up NVMe: {}".format(error))
    super(NvmeIo, self).start_daos()

And then tearDown() whould also be simpler:

def tearDown(self):
    try:
        super(NvmeIo, self).tearDown()
    finally:
        spdk.nvme(self.hostlist_servers, "cleanup")

Reply:
> It's a shame the `apricot.TestWithServers` class couldn't have the argument parsing section of it's `setUp()` be defined in one class method and the hostfile writing and agent/server startup in another class method,

It can.  It's fully intended to evolve as the needs for it change.  Until now there has not been a need to have those be separate but if the need arises, I completely endorse changing up the classes to fulfil new requirements.

Even without that change, this test could inherit `TestWithoutServers` and eliminate much of the above boilerplate.